File: packages/client/src/binary-doc-store.ts
import type { DocStoreEntry, JsonValue } from "@ktjn/searchable-format";
import { decodeDirectory } from "./binary-directory.js";
import { ByteReader } from "./byte-reader.js";

const STRUCTURED_MAGIC = new Uint8Array([0x53, 0x44, 0x4f, 0x43]);
const STRUCTURED_VERSION = 2;

/**
 * Client-side decoder for the directory-based binary doc store shard
 * encoding (`packages/indexer/src/binary-doc-store.ts`). Deliberately
 * lazy: `decodeBinaryDocStoreDirectory()` parses only the sorted
 * docId -> (byte offset, byte length) table, and
 * `decodeBinaryDocStoreEntry()` decodes exactly one document's stored
 * fields by seeking directly to its byte range -- a query only ever
 * needs the specific hit ids on its result page (typically a handful),
 * never every document in the (today, corpus-wide single) doc store
 * shard.
 */

export interface BinaryDocStoreDirectory {
  /** Every doc id in this shard, sorted ascending. */
  sortedIds: number[];
  /** docId -> byte offset/length into the records blob (relative to `directoryByteLength`). */
  index: Map<number, { offset: number; length: number }>;
  /** Byte offset where the records blob begins. */
  directoryByteLength: number;
}

/** Decodes just the directory (every doc id + its byte range) -- no record bytes are touched. */
export function decodeBinaryDocStoreDirectory(
  bytes: Uint8Array,
  binaryVersion?: number,
): BinaryDocStoreDirectory {
  const headerLength = structuredHeaderLength(bytes, binaryVersion);
  const r = new ByteReader(bytes, headerLength);
  const docCount = r.readVarint();
  let prevId = 0;
  const { keys: sortedIds, index } = decodeDirectory(r, docCount, (reader) => {
    prevId += reader.readVarint();
    return prevId;
  });
  const recordsLength = bytes.length - r.position;
  for (const { offset, length } of index.values()) {
    if (offset + length > recordsLength) {
      throw new Error("binary document-store record exceeds shard bounds");
    }
  }
  return { sortedIds, index, directoryByteLength: r.position };
}

/** Decodes exactly one document's record by seeking directly to its byte range. */
export function decodeBinaryDocStoreEntry(
  bytes: Uint8Array,
  directoryByteLength: number,
  offset: number,
  binaryVersion = 1,
): DocStoreEntry {
  const r = new ByteReader(bytes, directoryByteLength + offset);
  if (binaryVersion === 2) {
    assertMagic(r);
    const version = r.readVarint();
    if (version !== STRUCTURED_VERSION) {
      throw new Error(`unsupported binary document-store version: ${version}`);
    }
    const url = r.readString();
    const flags = r.readVarint();
    if ((flags & ~0x0f) !== 0) {
      throw new Error(`unsupported structured binary document flags: ${flags}`);
    }
    const boost = flags & 1 ? r.readFloat64() : undefined;
    const externalId = flags & 2 ? r.readString() : undefined;
    const contentHash = flags & 4 ? r.readString() : undefined;
    const metadata = flags & 8 ? readStructuredJsonValue(r) : undefined;
    const fields = readFields(r);
    const entry: DocStoreEntry = { url, fields };
    if (boost !== undefined) entry.boost = boost;
    if (externalId !== undefined) entry.externalId = externalId;
    if (contentHash !== undefined) entry.contentHash = contentHash;
    if (metadata !== undefined) entry.metadata = metadata;
    return entry;
  }
  if (binaryVersion !== 1) {
    throw new Error(
      `unsupported binary document-store version: ${binaryVersion}`,
    );
  }
  const url = r.readString();
  const hasBoost = r.readVarint() === 1;
  const boost = hasBoost ? r.readFloat64() : undefined;
  const fields = readFields(r);
  return boost !== undefined ? { url, boost, fields } : { url, fields };
}

function structuredHeaderLength(
  bytes: Uint8Array,
  binaryVersion?: number,
): number {
  const inferred = hasMagic(bytes) ? STRUCTURED_VERSION : 1;
  const version = binaryVersion ?? inferred;
  if (version === 1) return 0;
  if (version !== STRUCTURED_VERSION) {
    throw new Error(`unsupported binary document-store version: ${version}`);
  }
  const r = new ByteReader(bytes, 0);
  assertMagic(r);
  const encodedVersion = r.readVarint();
  if (encodedVersion !== STRUCTURED_VERSION) {
    throw new Error(
      `unsupported binary document-store version: ${encodedVersion}`,
    );
  }
  return r.position;
}

function hasMagic(bytes: Uint8Array): boolean {
  return STRUCTURED_MAGIC.every((value, index) => bytes[index] === value);
}

function assertMagic(reader: ByteReader): void {
  const magic = reader.readBytes(STRUCTURED_MAGIC.length);
  if (!hasMagic(magic)) {
    throw new Error("invalid structured binary document-store magic");
  }
}

function readFields(reader: ByteReader): Record<string, string> {
  const fieldCount = reader.readVarint();
  const fields: Record<string, string> = {};
  for (let i = 0; i < fieldCount; i++) {
    const fieldName = reader.readString();
    fields[fieldName] = reader.readString();
  }
  return fields;
}

function readStructuredJsonValue(reader: ByteReader): JsonValue {
  const tag = reader.readVarint();
  if (tag === 0) return null;
  if (tag === 1) return false;
  if (tag === 2) return true;
  if (tag === 3) {
    const value = reader.readFloat64();
    if (!Number.isFinite(value)) {
      throw new Error("structured binary metadata number must be finite");
    }
    return value;
  }
  if (tag === 4) return reader.readString();
  if (tag === 5) {
    return Array.from({ length: reader.readVarint() }, () =>
      readStructuredJsonValue(reader),
    );
  }
  if (tag === 6) {
    const object: Record<string, JsonValue> = {};
    const count = reader.readVarint();
    for (let i = 0; i < count; i++) {
      const key = reader.readString();
      if (Object.hasOwn(object, key)) {
        throw new Error(`duplicate structured binary metadata key: ${key}`);
      }
      object[key] = readStructuredJsonValue(reader);
    }
    return object;
  }
  throw new Error(`unsupported structured binary metadata tag: ${tag}`);
}

File: python/searchable-client/src/searchable_client/binary_doc_store.py
import math
from typing import Any

from searchable_client.byte_reader import ByteReader
from searchable_client.types import DocStoreEntry

_STRUCTURED_MAGIC = b"SDOC"
_STRUCTURED_VERSION = 2


def _read_structured_json_value(reader: ByteReader) -> Any:
    tag = reader.read_varint()
    if tag == 0:
        return None
    if tag == 1:
        return False
    if tag == 2:
        return True
    if tag == 3:
        value = reader.read_float64()
        if not math.isfinite(value):
            raise ValueError("structured binary metadata number must be finite")
        return value
    if tag == 4:
        return reader.read_string()
    if tag == 5:
        return [_read_structured_json_value(reader) for _ in range(reader.read_varint())]
    if tag == 6:
        result: dict[str, Any] = {}
        for _ in range(reader.read_varint()):
            key = reader.read_string()
            if key in result:
                raise ValueError(f"duplicate structured binary metadata key: {key}")
            result[key] = _read_structured_json_value(reader)
        return result
    raise ValueError(f"unsupported structured binary metadata tag: {tag}")


def _structured_header_length(data: bytes, binary_version: int | None) -> int:
    if binary_version is None:
        if not data.startswith(_STRUCTURED_MAGIC):
            return 0
        binary_version = _STRUCTURED_VERSION
    if binary_version == 1:
        return 0
    if binary_version != _STRUCTURED_VERSION:
        raise ValueError(f"unsupported binary document-store version: {binary_version}")
    if not data.startswith(_STRUCTURED_MAGIC):
        raise ValueError("invalid structured binary document-store magic")
    reader = ByteReader(data, len(_STRUCTURED_MAGIC))
    version = reader.read_varint()
    if version != _STRUCTURED_VERSION:
        raise ValueError(f"unsupported binary document-store version: {version}")
    return reader.position


def decode_binary_doc_store_directory(
    data: bytes,
    *,
    binary_version: int | None = None,
) -> tuple[list[int], dict[int, tuple[int, int]], int]:
    r = ByteReader(data, _structured_header_length(data, binary_version))
    doc_count = r.read_varint()
    sorted_ids: list[int] = []
    index: dict[int, tuple[int, int]] = {}
    prev_id = 0
    for _ in range(doc_count):
        prev_id += r.read_varint()
        doc_id = prev_id
        offset = r.read_varint()
        length = r.read_varint()
        sorted_ids.append(doc_id)
        index[doc_id] = (offset, length)
    records_length = len(data) - r.position
    if any(offset + length > records_length for offset, length in index.values()):
        raise ValueError("binary document-store record exceeds shard bounds")
    return sorted_ids, index, r.position


def decode_binary_doc_store_entry(
    data: bytes,
    directory_byte_length: int,
    offset: int,
    *,
    binary_version: int = 1,
) -> DocStoreEntry:
    r = ByteReader(data, directory_byte_length + offset)
    if binary_version == 2:
        if r.read_bytes(len(_STRUCTURED_MAGIC)) != _STRUCTURED_MAGIC:
            raise ValueError("invalid structured binary document-store magic")
        version = r.read_varint()
        if version != _STRUCTURED_VERSION:
            raise ValueError(f"unsupported binary document-store version: {version}")
        url = r.read_string()
        flags = r.read_varint()
        if flags & ~0x0F:
            raise ValueError(f"unsupported structured binary document flags: {flags}")
        boost = r.read_float64() if flags & 1 else None
        external_id = r.read_string() if flags & 2 else None
        content_hash = r.read_string() if flags & 4 else None
        metadata = _read_structured_json_value(r) if flags & 8 else None
        field_count = r.read_varint()
        structured_fields: dict[str, str] = {}
        for _ in range(field_count):
            field_name = r.read_string()
            structured_fields[field_name] = r.read_string()
        return DocStoreEntry(
            url=url,
            boost=boost,
            fields=structured_fields,
            external_id=external_id,
            metadata=metadata,
            content_hash=content_hash,
        )

    if binary_version != 1:
        raise ValueError(f"unsupported binary document-store version: {binary_version}")
    url = r.read_string()
    has_boost = r.read_varint() == 1
    boost = r.read_float64() if has_boost else None
    field_count = r.read_varint()
    fields: dict[str, str] = {}
    for _ in range(field_count):
        field_name = r.read_string()
        fields[field_name] = r.read_string()
    return DocStoreEntry(url=url, boost=boost, fields=fields)
